libtenzir/builtins/operators/grok.cpp

```diff
@@ -0,0 +1,518 @@
+          named_captures.erase(duplicate_it);
+        named_captures.emplace_back(name, conversion);
+        auto replacement
+          = fmt::format("(?<{}>{})", name, subpattern.resolved_pattern->str());
+        result_pattern.append(replacement);
+      } else {
+        // No NAME given, use SYNTAX as the name
+        if (auto duplicate_it = get_duplicate_capture_name(syntax);
+            duplicate_it != named_captures.end())
+          named_captures.erase(duplicate_it);
+        named_captures.emplace_back(syntax, capture_type::unnamed);
+        auto replacement = fmt::format("(?<{}>{})", syntax,
+                                       subpattern.resolved_pattern->str());
+        result_pattern.append(replacement);
+      }
+      // We'll also have all the same named captures as the subpattern,
+      // except if they have a name that we already have saved:
+      // we don't want subpattern captures to overwrite anything we have in the
+      // main pattern.
+      //
+      // We expect `named_captures` to be quite small, so O(n) lookup is okay
+      std::ranges::copy_if(subpattern.named_captures,
+                           std::back_inserter(named_captures),
+                           [&](const auto& capture) {
+                             return get_duplicate_capture_name(capture.first)
+                                    == named_captures.end();
+                           });
+      previous_match_end = replacement_field.end();
+    }
+    result_pattern.append(previous_match_end, str_re2.end());
+    resolved_pattern.emplace(std::move(result_pattern));
+    TENZIR_ASSERT(resolved_pattern);
+    TENZIR_ASSERT(not resolved_pattern->empty());
+  }
+}
+
+void pattern_store::parse_line(std::string_view line) {
+  if (line.empty())
+    return;
+  if (line.starts_with('#'))
+    return;
+  auto parts = detail::split(line, " ", 1);
+  TENZIR_ASSERT(not patterns.contains(std::string{parts[0]}));
+  patterns.emplace(std::string{parts[0]}, std::string{parts[1]});
+}
+
+auto& get_builtin_pattern_store() {
+  // Using CoW to avoid copying the builtin patterns unless we need to
+  static auto store = caf::make_copy_on_write<pattern_store>(
+    std::span{builtin_patterns_strings});
+  return store;
+}
+
+class grok_parser final : public plugin_parser {
+public:
+  grok_parser() = default;
+
+  explicit grok_parser(parser_interface& p)
+    : patterns_(get_builtin_pattern_store()) {
+    auto parser
+      = argument_parser{"grok", "https://docs.tenzir.com/next/operators/"
+                                "transformations/grok"};
+    parser.add(input_pattern_.raw_pattern, "<input_pattern>");
+    std::optional<std::string> pattern_definitions{};
+    parser.add("--pattern-definitions", pattern_definitions, "<patterns>");
+    parser.add("--indexed-captures", indexed_captures_);
+    parser.add("--include-unnamed", include_unnamed_);
+    parser.add("--include-nulls", include_nulls_);
```

Comment:
I'm not sure whether the `--include-nulls` option makes sense. Nulls would be included anyway if at least one of the strings in a batch matches the pattern. This is also one of the reasons why we generally want to treat `null` and absence the same.